sled-agent/src/metrics.rs

```diff
@@ -0,0 +1,177 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Metrics produced by the sled-agent for collection by oximeter.
+
+use oximeter::types::MetricsError;
+use oximeter::types::ProducerRegistry;
+use oximeter_instruments::kstat::link;
+use oximeter_instruments::kstat::CollectionDetails;
+use oximeter_instruments::kstat::Error as KstatError;
+use oximeter_instruments::kstat::KstatSampler;
+use oximeter_instruments::kstat::TargetId;
+use sled_hardware::Baseboard;
+use slog::Logger;
+use std::collections::BTreeMap;
+use std::time::Duration;
+use uuid::Uuid;
+
+/// The interval on which we ask `oximeter` to poll us for metric data.
+pub(crate) const METRIC_COLLECTION_INTERVAL: Duration = Duration::from_secs(30);
+
+/// The interval on which we sample link metrics.
+pub(crate) const LINK_SAMPLE_INTERVAL: Duration = Duration::from_secs(10);
```

Comment:
Are these values arbitrary, or is there a reasoning behind them? If it's the latter, it'd be nice to document it in the comments for later reference.

Reply:
Totally arbitrary right now. I'm trying to be pretty aggressive until we know that we need to reduce the data size in ClickHouse. I continue to be impressed with its compression ratios. I've left my system running over night, and it has accumulated about 80K samples from these new timeseries. Here is the size on disk:

```
oxz_clickhouse_3efa9966-379c-47b5-ad58-085609975eda :) select total_rows, total_bytes from system.tables where name = 'measurements_cumulativeu64';

SELECT
    total_rows,
    total_bytes
FROM system.tables
WHERE name = 'measurements_cumulativeu64'

Query id: bb566253-3b73-449f-8d96-4052511f3d50

┌─total_rows─┬─total_bytes─┐
│      79368 │      751315 │
└────────────┴─────────────┘

1 row in set. Elapsed: 0.002 sec.
```

So it's taking a bit fewer than 10 bytes per row. Walking that out, each timeseries should result in about 2.3MiB of data per month. At this point, I'm pretty happy paying that cost. We'll see how we feel as we add many more timeseries!

Reply:
Sweet! I am a little concerned about performance though. When we have a whole lot more metrics implemented, will ClickHouse get bogged down?

Reply:
Maybe? I am hoping to defer that until such time as we _do_ have many other metrics. Also, the point of picking ClickHouse was to minimize this possibility, and the design of our schema should make the query of any individual query scale very well with the number of timeseries. That's the goal anyway!

Reply:
Hmmm.. perhaps I am erring on the side of being too cautious. I'm a bit worried about the fact that we don't really have a good story around node recovery, other than ssh-ing into the zones to restart nodes, for now.

Would a more conservative sampling interval (perhaps per minute?) make sense for now? We can always increase the intervals down the track, but I believe customers may not like it if we decrease intervals once we ship metrics :smile: 

What do you think?

Reply:
I think that if we're concerned about resource usage, we should profile it using the introspection ClickHouse provides. I've given a rough estimate of disk usage for a 10s sample interval above, at around 2.3MiB per month. What about the consumption of other resources?

The memory usage will certainly depend on the queries, but as a reasonably pessimistic estimate, let's just access the entire timeseries:

```
0x〉select min(timestamp), max(timestamp), date_diff('h', min(timestamp), max(timestamp)) from physical_data_link:packets_sent;

 min(timestamp)                  | max(timestamp)                  | dateDiff('h', min(timestamp), max(timestamp))
---------------------------------+---------------------------------+-----------------------------------------------
 "2023-10-27 19:57:27.558080342" | "2023-10-28 18:52:36.857118267" | 23

Metadata
 Query ID:    81312a77-1ec3-4a92-987c-646dd976e1af
 Result rows: 1
 Time:        44.985431ms
 Read:        18469 rows (1080369 bytes)

```

That gives us an estimate for touching all rows for a 24 hour period. If we assume the performance scales linearly with the number of rows (not a great assumption, but not terrible), that would imply a little over 1s to run the same query on the full 30 days we will keep in the database right now. This is also using the prototype SQL shell I'm working on, which makes a huge join to reconstruct the timeseries as if it were a single table. So I expect that this query will be more resource-intensive than many others we might run, such as using the existing querying infrastructure which works very differently under the hood.

So suppose we're unhappy with the above performance, and we need to improve it. Can we do so? Yes. Going directly to the database to run the same query against the smallest slice of data needed to get the same answer, we see  this:

```
oxz_clickhouse_e449eb80-3371-40a6-a316-d6e64b039357 :) select min(timestamp), max(timestamp), date_diff('h', min(timestamp), max(timestamp)) from measurements_cumulativeu64 where timeseries_name = 'physical_data_link:packets_sent';

SELECT
    min(timestamp),
    max(timestamp),
    dateDiff('h', min(timestamp), max(timestamp))
FROM measurements_cumulativeu64
WHERE timeseries_name = 'physical_data_link:packets_sent'

Query id: be6fb311-71bf-4ff2-81c7-a4f74a8220e2

┌────────────────min(timestamp)─┬────────────────max(timestamp)─┬─dateDiff('h', min(timestamp), max(timestamp))─┐
│ 2023-10-27 19:57:27.558080342 │ 2023-10-28 18:56:36.899336796 │                                            23 │
└───────────────────────────────┴───────────────────────────────┴───────────────────────────────────────────────┘

1 row in set. Elapsed: 0.006 sec. Processed 17.34 thousand rows, 835.38 KB (3.05 million rows/s., 146.70 MB/s.)

``` 

I.e., it takes 6ms to answer this question for the data I've accumulated over around 24 hours. This touches all data for the timeseries, and ClickHouse absolutely blazes through it.

As I mentioned above, because of how we've written the indexes, adding another timeseries, say `physical_data_link:bytes_sent`, has no impact on this query. The system can and does use its sorting key to skip to the correct part of the table:

```
oxz_clickhouse_e449eb80-3371-40a6-a316-d6e64b039357 :) explain indexes=1 select min(timestamp), max(timestamp), date_diff('h', min(timestamp), max(timestamp)) from measurements_cumulativeu64 where timeseries_name = 'physical_data_link:packets_sent';

EXPLAIN indexes = 1
SELECT
    min(timestamp),
    max(timestamp),
    dateDiff('h', min(timestamp), max(timestamp))
FROM measurements_cumulativeu64
WHERE timeseries_name = 'physical_data_link:packets_sent'

Query id: 9a94bb05-931d-41c9-91de-4bd359cc6d7d

┌─explain────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))                                                                        │
│   Aggregating                                                                                                      │
│     Expression (Before GROUP BY)                                                                                   │
│       Filter (WHERE)                                                                                               │
│         ReadFromMergeTree (oximeter.measurements_cumulativeu64)                                                    │
│         Indexes:                                                                                                   │
│           PrimaryKey                                                                                               │
│             Keys:                                                                                                  │
│               timeseries_name                                                                                      │
│             Condition: (timeseries_name in ['physical_data_link:packets_sent', 'physical_data_link:packets_sent']) │
│             Parts: 5/5                                                                                             │
│             Granules: 7/17                                                                                         │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

12 rows in set. Elapsed: 0.003 sec.

```

There are actually 37 parts in this table at the time I ran this query:

```
oxz_clickhouse_e449eb80-3371-40a6-a316-d6e64b039357 :) select count() from system.parts where table = 'measurements_cumulativeu64';

SELECT count()
FROM system.parts
WHERE table = 'measurements_cumulativeu64'

Query id: 4ac7bf06-2050-4b42-af94-ad8f95e6cf00

┌─count()─┐
│      37 │
└─────────┘

1 row in set. Elapsed: 0.002 sec.

```

Overall, my judgement here is that the evidence suggests ClickHouse will perform quite well even at such high sampling rates. At this time, I am willing to trade the improved observability of our own system against an uncertain and future risk of customer dissatisfaction.

Reply:
> Hmmm.. perhaps I am erring on the side of being too cautious. I'm a bit worried about the fact that we don't really have a good story around node recovery, other than ssh-ing into the zones to restart nodes, for now.
> 
> Would a more conservative sampling interval (perhaps per minute?) make sense for now? We can always increase the intervals down the track, but I believe customers may not like it if we decrease intervals once we ship metrics 😄

Over time, we need to get a point where we're realistically going to be gathering these on per-second basis to get useful information and be actionable. A 1 minute average is not useful, even 10s averages aren't great. However, just because we're gathering it at a given rate also doesn't mean that the API has to expose the same information per se if we want to worry about breaking customer expectations. Honestly, even in networking there are microbursts that occur where the 1s bandwidth shows that you aren't saturated, but the reality is that for a given 10ms period we were which causes congestion and packet loss. I don't mean that to suggest that we're ever going to get to a 10ms sampling period or that it makes sense, but just that long averages don't actually help us. This isn't specific to the network, it's true for power utilization and more. Temperature is something that is harder to change just due to the nature of physics so it's going to be a trailing indicator. There are also other ways to phrase and capture data that can ameliorate this for gauges like temperature or current, but there is no such amelioration for counters.

I think Ben has demonstrated initial PoCs that show that this is pretty reasonable. However, if we're going to continue being conservative, that can only be accompanied with an actionable plan for how we would measure and build the confidence. That plan can't really rely on waiting for deployments or customers, but needs to be something that we're proactively taking action on.